### DISCUSSION TITLE
[LGTM] C/C++ build system undetected for Magma

### DISCUSSION THREAD

**Discussion by bbarritt** [github/securitylab#256] [Q&A] [answered]
We're new to LGTM but are very excited to have enabled it on our open source project for [Magma](https://www.magmacore.org/).  It's finding really useful insights on our js/ts, Go, and Python code; but, it's currently unable to analyze the C/C++ in the same repo.  We use regular Makefile builds and care mostly about it analyzing the code in the ./lte/gateway/ directory.  But it seems to not detect the build system at all.  It's unclear to me how to resolve using a project configuration file, since there are no special configuration steps required.  Can you please advise?

Note that our project does require a number of libraries to be installed for openvswitch, but I don't think we're even getting that far.  You can find our proejct here: 
https://lgtm.com/projects/g/magma/magma/logs/languages/lang:cpp

I've also attached our extraction log.
[dd6368d8e5fb87d7c503cf199a8b8f4b886cd728_build.log](https://github.com/github/securitylab/files/5890075/dd6368d8e5fb87d7c503cf199a8b8f4b886cd728_build.log)

**Comment by jbj** [comment, marked as answer]
The build is failing because LGTM wasn't able to guess that `lte/gateway` is the interesting directory. There are two ways you can provide that information
1. Create a `.lgtm.yml` file and [customize the `index` step](https://lgtm.com/help/lgtm/cpp-extraction) with something like `build_command: make -C lte/gateway`.
1. Give the project a top-level `Makefile` that's a wrapper for `lte/gateway` and any other directories that should be compiled by default. Then you're also providing build information to humans and other automated build systems.

Dependencies that exist as Ubuntu packages should be installed automatically. If that fails for some reason, the `.lgtm.yml` allows specifying them manually.

**Reply by bbarritt** [reply]
I tried this and LGTM was able to find the Makefile, but the build is failing for other reasons. It turns out that as of right now, our build environment is highly customized and relies upon a packet script to build a base Debian VM and Ansible playbooks to set up the remaining build environment. Our team is working to simplify the build process in our next release, but that's several months away.

Is there a way to configure LGTM to find and analyze C/C++ code based on a directory target or other mechanism without actually having to build the code? I could imagine that such a feature might be useful for projects, too (especially those that can't build code from Linux).

**Reply by jbj** [reply]
For C/C++, LGTM can only analyze code that it can build. The LGTM build environment is not very flexible; at the moment it's Ubuntu 19.10 inside a Docker container.

If you can build your code using GitHub Actions, you can [set up Code Scanning](https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/enabling-code-scanning-for-a-repository). Code Scanning uses the same underlying technology as LGTM, called CodeQL, but is integrated with GitHub.
